K8s bootstrap docs

chbatey opened the pull request:
I'll update the demon as well but thought it is worth getting @TimMoore @raboof @ktoso thoughts on this so I don't need to change it again :)

chbatey commented on docs/src/main/paradox/bootstrap/kubernetes.md:
```diff
@@ -0,0 +1,149 @@
+  selector:
+    appName: "akka-cluster-kubernetes"
+  clusterIP: None
+  publishNotReadyAddresses: true
```
@TimMoore this is the property for publishing DNS entries for not ready

ktoso replied:
I see, nice! Where's that in docs?

ktoso replied:
Thought that's an url but isn't hm http://service.alpha.kubernetes.io/tolerate-unready-endpoints

chbatey replied:
Only in the swagger for the api server from what I can see. 

I found it via its github issue.

chbatey commented on docs/src/main/paradox/bootstrap/kubernetes.md:
```diff
@@ -0,0 +1,149 @@
+      port: 8558
+      protocol: TCP
+      targetPort: 8558
+    - name: remoting
```
this doesn't actually need to be on here as the remoting port is discovered via bootstrap

ktoso commented on docs/src/main/paradox/bootstrap/kubernetes.md:
````diff
@@ -0,0 +1,149 @@
+```
+
+Note there are currently two ways to set that addressed should be published if not ready, the initial way via an annotation 
+`service.alpha.kubernetes.io/tolerate-unready-endpoints` and via the new officially supported way as the property `publishNotReadyAddresses`.
````
link to its docs?

chbatey replied:
Ah but that would require it to be documented :)

ktoso commented on docs/src/main/paradox/bootstrap/kubernetes.md:
```diff
@@ -0,0 +1,149 @@
+  labels:
+    appName: "akka-cluster-kubernetes"
+  annotations:
+    service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
```
really "true" or true would work too? for my info

chbatey replied:
Annotations have to be strings :(

If you have it as true, `kubectl create` tells you it is invalid and `kubectl apply` ignores it so it doesn't get set (reported bug)

ktoso commented on docs/src/main/paradox/bootstrap/kubernetes.md:
```diff
@@ -0,0 +1,149 @@
+        if (readyStates.contains(selfState)) complete(StatusCodes.OK)
+        else complete(StatusCodes.InternalServerError)
+      }
+    } ~
```
please use `concat()`

raboof commented on docs/src/main/paradox/bootstrap/kubernetes.md:
```diff
@@ -0,0 +1,149 @@
+
+* Use Akka Bootstrap with `akka-dns` with cluster formation via DNS SRV records
+* Use a headless service for internal for Akka management/bootstrap so that readiness probes for prod traffic don't interfere with bootstrap 
+* If required use a separate service and/or ingress for user facing endpoints e.g. HTTP gRPC
```
Maybe add some more punctiation, like "ingress for user-facing endpoints, for example HTTP or gRPC"? Maybe add links to akka-http and akka-grpc? Do we know if ingress works for gRPC?

chbatey replied:
I haven't tried ingress for gRPC only internal comms. It'll depend on what is your ingress controller, most of them are nginx

raboof commented on docs/src/main/paradox/bootstrap/kubernetes.md:
````diff
@@ -0,0 +1,149 @@
+  publishNotReadyAddresses: true
+```
+
+Note there are currently two ways to set that addressed should be published if not ready, the initial way via an annotation 
````
maybe 'specify' rather than 'set'?

raboof commented on docs/src/main/paradox/bootstrap/kubernetes.md:
```diff
@@ -0,0 +1,149 @@
+
+Note there are currently two ways to set that addressed should be published if not ready, the initial way via an annotation 
+`service.alpha.kubernetes.io/tolerate-unready-endpoints` and via the new officially supported way as the property `publishNotReadyAddresses`.
+Set both as depending on your DNS solution it may have not migrated from the annotation to the property.
```
Maybe record which versions of common DNS solutions are updated to understand the 'new' way?

raboof commented on docs/src/main/paradox/bootstrap/kubernetes.md:
```diff
@@ -0,0 +1,149 @@
+
+The same configuration will work for any environment that has an SRV record for your Akka Clustered application. 
+
+For prod traffic e.g. HTTP use a regular service or an alternative ingress mechanism. 
```
Maybe add 2 subsection headers in this section ('Headless service' and 'Exposed service'?)

chbatey replied:
:+1:

raboof commented on docs/src/main/paradox/bootstrap/kubernetes.md:
```diff
@@ -0,0 +1,149 @@
+The same configuration will work for any environment that has an SRV record for your Akka Clustered application. 
+
+For prod traffic e.g. HTTP use a regular service or an alternative ingress mechanism. 
+With an appropriate readiness check this results in traffic not being routed until bootstrap has finished.
```
:+1:

chbatey commented:
This won't build until 2.5.15 is released.

I've updated the docs to take the yaml / config from a working demo that i've added to this project.

I've also restructured the demos so they don't get up with such long names when building docker images for them

chbatey commented on project/Dependencies.scala:
```diff
@@ -3,7 +3,7 @@ import Keys._
 object Dependencies {
 
-  val AkkaVersion = "2.5.14"
+  val AkkaVersion = "2.5-SNAPSHOT"
```
Will update this to 2.5.15 when released

raboof commented on bootstrap-demo/kubernetes-dns/kubernetes/akka-cluster.yml:
```diff
@@ -0,0 +1,82 @@
+metadata:
+  labels:
+    appName: "bootstrap-demo-kubernetes-dns"
+  name: "bootstrap-demo-kubernetes-dns-public"
```
maybe remove the `-public` suffix here and add a `-headless` or `-akka-internal` or something similar to the headless service?

chbatey replied:
Yep good idea

markusjura commented:
Very nice solution and writeup 👍

raboof commented on docs/src/main/paradox/bootstrap/demos/kubernetes.md:
```diff
@@ -0,0 +1,75 @@
+
+## Services
+
+## Internal headless service
```
Should be a `###` heading?

raboof commented on docs/src/main/paradox/bootstrap/demos/kubernetes.md:
```diff
@@ -0,0 +1,75 @@
+
+### Update strategy
+
+For small clusters it may make sense to set `maxUnabailable` to 0 and `maxSurge` to 1. 
```
typo

raboof commented on docs/src/main/paradox/bootstrap/demos/kubernetes.md:
```diff
@@ -0,0 +1,75 @@
+
+## Deployments
+
+Use a regular deployment (not a StatefulSet). 
```
Perhaps explain why?

raboof commented on docs/src/main/paradox/bootstrap/demos/kubernetes.md:
```diff
@@ -0,0 +1,75 @@
+
+@@snip [health-checks]($management$/bootstrap-demo/kubernetes-dns/src/main/scala/akka/cluster/bootstrap/KubernetesHealthChecks.scala)  { #health }
+
+This will mean that a pod won't get traffic until it is part of a cluster which is important
```
Maybe add a comma, 'until it is part of a cluster, which is ...'

raboof commented on docs/src/main/paradox/bootstrap/demos/kubernetes.md:
```diff
@@ -0,0 +1,75 @@
+# Kubernetes
+
+A full working example that can be deployed to `minikube` is in `bootstrap-demo/kuvernetes-dns`.
```
kuvernetes, va?

raboof commented on bootstrap-demo/kubernetes-dns/src/main/scala/akka/cluster/bootstrap/KubernetesHealthChecks.scala:
```diff
@@ -0,0 +1,42 @@
+  //#health
+  private val readyStates: Set[MemberStatus] = Set(MemberStatus.Up, MemberStatus.WeaklyUp)
+  private val aliveStates: Set[MemberStatus] =
+    Set(MemberStatus.Joining, MemberStatus.WeaklyUp, MemberStatus.Up, MemberStatus.Leaving, MemberStatus.Exiting)
```
Hm, indeed we don't want K8s to kill the container while it is leaving/exiting, OTOH we want K8s to start removing the node from the loadbalancer, since otherwise there could be a window in which it still might get traffic while it has been moved to `Down`, right?

When a node is no longer 'alive', shouldn't K8s keep in account a grace period, and even when it decides to kill the node, keep a healthy delay between `SIGTERM` and `SIGKILL`?

chbatey replied:
I believe that is what readiness is for, k8s keeps hitting the readiness probe which will only succeeds in Up and WeeklyUp. So we can be alive but not ready when shutting down

raboof replied:
Ah yes I incorrectly assumed 'ready' would stop polling as soon as the pod has become ready. I agree this is :+1: then - we'd only need to make sure nodes aren't downed faster than the `readinessProbe` `periodSeconds` plus some buffer.

chbatey commented on bootstrap-demo/kubernetes-dns/kubernetes/akka-cluster.yml:
```diff
@@ -0,0 +1,88 @@
+    protocol: TCP
+    targetPort: 8080
+  selector:
+    appName: "bootstrap-demo-kubernetes-dns"
```
:+1:

TimMoore commented:
Sorry for the delay. Looks good to me, too. I think it's OK to go ahead and merge.

I'll also summon @eed3si9n, @cunei, @renatocaval and @dwijnand so they're aware of it. It's relevant to what they're working on.

TimMoore commented on docs/src/main/paradox/bootstrap/demos/kubernetes-api.md:
```diff
@@ -54,7 +55,7 @@ $ sbt shell
+@@snip [kube-create.sh]($management$/bootstrap-demo/kubernetes-api/kube-create.sh) 
 
 This will create and start running a number of Pods hosting the application. The application nodes will proceed with 
 forming the cluster using the DNS bootstrap method. In order to observe the logs during the cluster formation you can 
```
Just noticed that this page talks about the DNS bootstrap method, but it's supposed to be the Kubernetes API demo.

chbatey replied:
Raised  https://github.com/akka/akka-management/issues/292 to track making that example purely k8s API

TimMoore commented on docs/src/main/paradox/bootstrap/demos/kubernetes.md:
```diff
@@ -0,0 +1,75 @@
+# Kubernetes
+
+A full working example that can be deployed to kubernetes via `minikube` is in `bootstrap-demo/kuvernetes-dns`.
```
small typo: kuvernetes

TimMoore commented on docs/src/main/paradox/bootstrap/demos/index.md:
```diff
@@ -0,0 +1,4 @@
+# Bootstrap examples 
```
This page could be hard to find, because it's not listed at the top-level menu, only linked from the bottom of the bootstrap page.

TimMoore commented on docs/src/main/paradox/bootstrap/demos/kubernetes.md:
```diff
@@ -0,0 +1,75 @@
+
+A full working example that can be deployed to kubernetes via `minikube` is in `bootstrap-demo/kuvernetes-dns`.
+
+As of Akka 2.5.15 and Akka-Management 0.18 the recommended way to run Akka Cluster in Kubernetes is to:
```
Maybe update to 2.5.16 now

longmuir commented on bootstrap-demo/kubernetes-dns/kubernetes/akka-cluster.yml:
```diff
@@ -0,0 +1,89 @@
+  labels:
+    appName: "bootstrap-demo-kubernetes-dns"
+  annotations:
+    service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
```
It looks like this annotation will be removed in Kubernetes 1.13 (https://github.com/kubernetes/kubernetes/issues/63741) to be replaced by `Service.spec.publishNotReadyAddresses` (https://github.com/kubernetes/kubernetes/commit/944e07480fa4066fb1345cd1a81d74b3bbb78440). As a general rule, I don't think we should be telling people to use Alpha K8s features. Should we wait until this is sorted out in K8s before releasing?

raboof replied:
In the docs at https://github.com/akka/akka-management/pull/278/files#diff-e737933b9744b8db3a71650e02a1e627R50 we recommend adding both to avoid problems with different versions/assumptions - that seems reasonably safe?

ktoso replied:
I don't think we know; would have to be looked into. I'll mark it as an action item though.
@chbatey is now laser focused on shipping another thing for a customer, we'll get his attention later on to this.

longmuir commented:
See note about `service.alpha.kubernetes.io/tolerate-unready-endpoints annotation` being deprecated and shortly removed. I'm wondering if we should hold off on this until Kubernetes 1.13 is released with https://github.com/kubernetes/kubernetes/commit/944e07480fa4066fb1345cd1a81d74b3bbb78440. The challenge with recommending this, is that when the change happens each individual user/customer will have to adjust their K8s configs, and it may not be clear why failures are happening. I'm also not clear if this annotation is supported on OpenShift.

There may have to be different recommendations depending on the version or packaging of K8s that people are using.

chbatey commented:
It’s an alpha feature which has made it into main line

On Fri, 31 Aug 2018 at 15:20, Jamie Longmuir <notifications@github.com>
wrote:

> *@longmuir* commented on this pull request.
> ------------------------------
>
> In bootstrap-demo/kubernetes-dns/kubernetes/akka-cluster.yml
> <https://github.com/akka/akka-management/pull/278#discussion_r214369373>:
>
> > +        ports:
> +        - containerPort: 8558
> +          name: management
> +        - containerPort: 2552
> +          name: remoting
> +        volumeMounts: []
> +      volumes: []
> +---
> +#headless
> +apiVersion: v1
> +kind: Service
> +metadata:
> +  labels:
> +    appName: "bootstrap-demo-kubernetes-dns"
> +  annotations:
> +    service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
>
> Do we know if one/both of these are available on OpenShift and other cloud
> hosted K8s providers? I can't find anything about the alpha annotation on
> OpenShift, but only searched briefly. As a general rule, I think we should
> avoid recommending Alpha features as they're likely to stay in production.
> I think it's important that people know that the alpha annotation *will*
> be removed and that the new one will be added in case there are problems
> when that happens. K8s releases have not always been smooth.
>
> Alternatively, we could label this solution for Cluster formation as
> experimental until the annotations sort themselves. Or at least validate
> that this works/does not work on OpenShift, GKE, AKS, etc. There's a lot
> riding on this functionality.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/akka/akka-management/pull/278#discussion_r214369373>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABx8G6gI3CQad9Jo04EJ5PI0Nys1QQB1ks5uWUZDgaJpZM4WHhyJ>
> .
>

raboof commented:
I think @longmuir makes some great points above and we should definitely follow up and test some of our assumptions.

On the other hand I think the content in the PR could already be extremely valuable for those situations where this annotation/setting is available.

Should we perhaps rephrase some of the introduction here, warn that it is not widely tested and that this relatively-new feature is required, and merge this PR relatively quickly but without immediately elevating this approach to 'The Recommended Way'?

chbatey commented:
+1 I’ll update this next week as saying this is one way

On Fri, 7 Sep 2018 at 13:45, Arnout Engelen <notifications@github.com>
wrote:

> I think @longmuir <https://github.com/longmuir> makes some great points
> above and we should definitely follow up and test some of our assumptions.
>
> On the other hand I think the content in the PR could already be extremely
> valuable for those situations where this annotation/setting is available.
>
> Should we perhaps rephrase some of the introduction here, warn that it is
> not widely tested and that this relatively-new feature is required, and
> merge this PR relatively quickly but without immediately elevating this
> approach to 'The Recommended Way'?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/akka/akka-management/pull/278#issuecomment-419427963>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABx8G9rXGKVpDzTWGgl0OBDT3ZJdryoFks5uYmppgaJpZM4WHhyJ>
> .
>

chbatey commented:
Updated saying they are alternatives. I think the kubernetes api demo needs some work as it mentions dns created https://github.com/akka/akka-management/issues/292 to follow up
